src/libfuncs/felt252_dict_entry.rs

```diff
@@ -106,61 +121,78 @@ pub fn build_get<'ctx, 'this>(
+            let block = region.append_block(Block::new(&[]));
+            let helper = LibfuncHelper {
+                module: helper.module,
+                init_block: helper.init_block,
+                region: &region,
+                blocks_arena: helper.blocks_arena,
+                last_block: std::cell::Cell::new(&block),
+                branches: vec![],
+                results: vec![],
+            };
+
+            let value = registry
+                .get_type(&info.branch_signatures()[0].vars[1].ty)?
+                .build_default(
                     context,
-                    block_occupied,
+                    registry,
+                    &block,
                     location,
-                    &info.ty,
-                    value,
-                )?
-            }
-            _ => (value, value),
-        };
+                    &helper,
+                    metadata,
+                    &info.branch_signatures()[0].vars[1].ty,
+                )?;
+            block.append_operation(scf::r#yield(&[value], location));
 
-        block_occupied.store(context, location, entry_value_ptr, values.0)?;
-        block_occupied.append_operation(cf::br(block_final, &[values.1], location));
-    }
+            region
+        },
+        {
+            // If the entry is occupied, then load the previous value from the pointer
+            let region = Region::new();
+            let block = region.append_block(Block::new(&[]));
 
-    {
-        let value = registry
-            .get_type(&info.branch_signatures()[0].vars[1].ty)?
-            .build_default(
-                context,
-                registry,
-                block_vacant,
-                location,
-                helper,
-                metadata,
-                &info.branch_signatures()[0].vars[1].ty,
-            )?;
-        block_vacant.append_operation(cf::br(block_final, &[value], location));
-    }
+            let value = block.load(context, location, entry_value_ptr, value_ty)?;
+
+            block.append_operation(scf::r#yield(&[value], location));
+
+            region
+        },
+        location,
+    ))?;
+
+    // Build the dict entry struct: `crate::types::felt252_dict_entry`
+    let dict_entry = entry.append_op_result(llvm::undef(entry_ty, location))?;
+    let dict_entry = entry.insert_values(
+        context,
+        location,
+        dict_entry,
+        &[dict_ptr, entry_value_ptr_ptr],
+    )?;
 
-    let entry = block_final.append_op_result(llvm::undef(entry_ty, location))?;
-    let entry =
-        block_final.insert_values(context, location, entry, &[dict_ptr, entry_value_ptr_ptr])?;
+    // A this point, the internal state holds two references to the current dict
+    // entry value until the finalize libfunc is called. If the compiler were to drop
+    // the dict entry as well as the value, then there would be undefined behaviour.
+    // This should never happen, as a dict entry must always be finalized.
```

Comment:
```suggestion
    // The `Felt252DictEntry<T>` holds both the `Felt252Dict<T>` and the pointer
    // to the space where the new value will be written when the entry is finalized.
    // If the entry were to be dropped (without being consumed by the finalizer), which
    // shouldn't be possible under normal conditions, and the type `T` requires a
    // custom drop implementation (ex. arrays, dictionaries), it'll cause undefined
    // behavior because when the value is moved out of the dictionary (on `get`), the
    // memory it occupied is not modified because we're expecting it to be overwritten
    // by the finalizer (in other words, the extracted element will be dropped twice).
```